src/qibo/models.py

```diff
@@ -198,70 +199,178 @@ def loss(params):
+            If ``None``, :class:`qibo.hamiltonians.X` is used.
+        solver (str): solver used to apply the exponential operators.
+        callbacks (list): List of callbacks to calculate during evolution.
+        accelerators (dict): Dictionary of devices to use for distributed
+            execution. See :class:`qibo.tensorflow.distcircuit.TensorflowDistributedCircuit`
+            for more details. This option is available only when ``hamiltonian``
+            is a :class:`qibo.base.hamiltonians.TrotterHamiltonian`.
+        memory_device (str): Name of device where the full state will be saved.
+            Relevant only for distributed execution (when ``accelerators`` is
+            given).
+
+    Example:
+        ::
+
+            import numpy as np
+            from qibo import models, hamiltonians
+            # create XXZ Hamiltonian for four qubits
+            hamiltonian = hamiltonians.XXZ(4)
+            # create QAOA model for this Hamiltonian
+            qaoa = models.QAOA(hamiltonian)
+            # optimize using random initial variational parameters
+            # and default options and initial state
+            initial_parameters = 0.01 * np.random.random(4)
+            best_energy, final_parameters = qaoa.minimize(initial_parameters, method="BFGS")
+    """
+    from qibo import hamiltonians, optimizers
+    from qibo.config import K, DTYPES, log
+    from qibo.callbacks import Norm
+
+    def __init__(self, hamiltonian, mixer=None, solver="exp", callbacks=[],
+                 accelerators=None, memory_device="/CPU:0"):
+        # list of QAOA variational parameters (angles)
+        self.params = None
         # problem hamiltonian
         self.hamiltonian = hamiltonian
+        self.nqubits = hamiltonian.nqubits
         # mixer hamiltonian (default = -sum(sigma_x))
-        if mixer == None:
-            self.mixer = self.hamiltonians.X(nqubits)
+        if mixer is None:
+            trotter = isinstance(
+                self.hamiltonian, self.hamiltonians.TrotterHamiltonian)
+            self.mixer = self.hamiltonians.X(self.nqubits, trotter=trotter)
         else:
+            if type(mixer) != type(hamiltonian):
+                  raise_error(TypeError, "Given Hamiltonian is of type {} "
+                                         "while mixer is of type {}."
+                                         "".format(type(hamiltonian),
+                                                   type(mixer)))
             self.mixer = mixer
-        # initial state
-        if initial_state == None:
-            self.initial_state = self.np.ones(2**nqubits) / self.np.sqrt(2**nqubits)
-        else:
-            self.initial_state = initial_state
-                   
-    def loss(self, p, solver='exp'):
+
+        # create circuits for Trotter Hamiltonians
+        if (accelerators is not None and (
+            not isinstance(self.hamiltonian, self.hamiltonians.TrotterHamiltonian)
+            or solver != "exp")):
```

Comment:
Ok, this means that rk is not supported, is there a technical motivation for that?

Reply:
Maybe that since `rk` is not unitary, then you couldn't really map it to an actual quantum circuit executed on QPU (at least in a direct and straightforward manner)?

Reply:
RK is not supported only when distributed execution is used, this `NotImplemetedError` is raised only when RK is used and `accelerators` is not `None`. For single device execution RK is supported and working.
The reason is what Diego said, distributed execution uses the `DistributedCircuit` class, so it requires the evolution/QAOA to be mapped to a circuit. In the current implementation this is possible only when `TrotterHamiltonian` + "exp" solver is used. RK is indeed not unitary so it is not possible to map it to a circuit anyway (regardless of implementation).

Note that it is possible to use RK + `TrotterHamiltonian`. This will calculate the dot products between Hamiltonian and state term-by-term using two-qubit gates, instead of creating the full Hamiltonian matrix. This is still not unitary because the various terms are summed (and also requires copies of the state vector).

Reply:
Ok, that's clear, thanks.